ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -143,6 +148,126 @@ object ContractDao {
+    *
+    * It is perfectly fine for an offset to be returned but the set of lagging template IDs
+    * be empty; this means they appear to be consistent but not at `expectedOffset`; since
+    * that means the state at query time is indeterminate, the query must be rerun anyway,
+    * but no further DB updates are needed.
+    *
+    * @param parties The party-set that must not be lagging.  We aren't concerned with
+    *          whether ''other'' parties are lagging, in fact we can't correct if they are,
+    *          but if another party got ahead, that means one of our parties lags.
+    * @param expectedOffset `fetchAndPersist` should have synchronized every template-ID/party
+    *          pair to the same offset, this one.
+    * @param templateIds The template IDs we're checking.
+    * @return Any template IDs that are lagging, and the offset to catch them up to, if defined;
+    *         otherwise everything is fine.
+    */
+  def laggingOffsets(
+      parties: Set[domain.Party],
+      expectedOffset: domain.Offset,
+      templateIds: NonEmpty[Set[domain.TemplateId.RequiredPkg]],
+  )(implicit
+      log: LogHandler,
+      sjd: SupportedJdbcDriver.TC,
+      lc: LoggingContextOf[InstanceUUID],
+  ): ConnectionIO[Option[(domain.Offset, Set[domain.TemplateId.RequiredPkg])]] = {
+    type Unsynced[Party, Off] = Map[Queries.SurrogateTpId, Map[Party, Off]]
+    import scalaz.syntax.traverse._
+    import sjd.q.queries.unsyncedOffsets
+    for {
+      tpids <- {
+        import Queries.CompatImplicits.monadFromCatsMonad
+        templateIds.toVector.toF.traverse { trp => surrogateTemplateId(trp) map ((_, trp)) }
+      }: ConnectionIO[NonEmptyF[Vector, (SurrogateTpId, domain.TemplateId.RequiredPkg)]]
+      surrogatesToDomains = tpids.toMap
+      unsyncedRaw <- unsyncedOffsets(
+        domain.Offset unwrap expectedOffset,
+        surrogatesToDomains.keySet,
+      )
+      unsynced = domain.Party.subst[Unsynced[*, domain.Offset], String](
+        domain.Offset.tag.subst[Unsynced[String, *], String](unsyncedRaw)
+      ): Unsynced[domain.Party, domain.Offset]
+    } yield minimumViableOffsets(parties, surrogatesToDomains, expectedOffset, unsynced)
+  }
+
+  // postprocess the output of unsyncedOffsets
+  private[dbbackend] def minimumViableOffsets[ITpId, OTpId, Party, Off: Order](
+      queriedParty: Set[Party],
+      surrogatesToDomains: ITpId => OTpId,
+      expectedOffset: Off,
+      unsynced: Map[ITpId, Map[Party, Off]],
+  ): Option[(Off, Set[OTpId])] = {
+    import scalaz.syntax.foldable1.{ToFoldableOps => _, _}
+    import scalaz.std.iterable._
+    val lagging: Option[Lagginess[ITpId, Off]] =
+      (unsynced: Iterable[(ITpId, Map[Party, Off])]) foldMap1Opt {
+        case (surrogateTpId, partyOffs) =>
+          val maxLocalOff = partyOffs
+            .collect {
+              case (unsyncedParty, unsyncedOff)
+                  if queriedParty(unsyncedParty) || unsyncedOff > expectedOffset =>
```

Comment:
what should be the behaviour for an `unsyncedParty` which is not a `queriedParty` and has `unsyncedOff < expectedOffset` ?

Reply:
We can safely ignore it. This is the test `"not require update for lagging, but unqueried, party"`.